distributed/protocol/core.py

```diff
@@ -45,31 +48,54 @@ def _inplace_compress_frames(header, frames):
 
             header["compression"] = tuple(compression)
 
+        def create_sub_frames(obj) -> list:
+            typ = type(obj)
+            if typ is Serialized:
+                sub_header, sub_frames = obj.header, obj.frames
+            else:
+                sub_header, sub_frames = serialize_and_split(
+                    obj,
+                    serializers=serializers,
+                    on_error=on_error,
+                    context=context,
+                    size=frame_split_size,
+                )
+                _inplace_compress_frames(sub_header, sub_frames)
+            sub_header["num-sub-frames"] = len(sub_frames)
+            sub_header = msgpack.dumps(
+                sub_header, default=msgpack_encode_default, use_bin_type=True
+            )
+            return [sub_header] + sub_frames
+
+        def create_pickle_sub_frames(obj) -> list:
+            typ = type(obj)
+            if typ is Pickled:
+                sub_header, sub_frames = obj.header, obj.frames
+            else:
+                sub_frames = []
+                sub_header = {
+                    "pickled-obj": pickle.dumps(
+                        obj.data, buffer_callback=sub_frames.append
+                    )
+                }
+                _inplace_compress_frames(sub_header, sub_frames)
```

Comment:
It seems that `sub_frames` can sometimes contain `pickle.PickleBuffer` objects (which have no `__len__` attribute). This seems to cause errors in `maybe_compress`.

Perhaps something like this makes sense?:

```python
                _inplace_compress_frames(
                    sub_header,
                    [sf.raw() if hasattr(sf, "raw") else sf for sf in sub_frames],
                )
```

Reply:
You are right, I am now using `ensure_memoryview()` to make all out-of-band buffers regular memoryview of datatype bytes:
```python
  # In to support len() and slicing, we convert `PickleBuffer`
  # objects to memoryviews of bytes.
  buffer_callback=lambda x: sub_frames.append(
      ensure_memoryview(x)
  ),
```